### PR TITLE
#1129 Do not use equals and hashCode from TypeMirror

### LINKED ISSUE #1129: Don't use equals and hashCode from TypeMirror
This is connected to the fix to #1127. There is [this](https://bugs.openjdk.java.net/browse/JDK-7159498) JDK bug report that might be causing java lang `ClassType` that are actually same to be different. 

According to the javadoc of [`TypeMirror#equals()`](https://docs.oracle.com/javase/8/docs/api/javax/lang/model/type/TypeMirror.html#equals-java.lang.Object-) one should use [`Types#isSameType()`](https://docs.oracle.com/javase/8/docs/api/javax/lang/model/util/Types.html#isSameType-javax.lang.model.type.TypeMirror-javax.lang.model.type.TypeMirror-) for semantic comparison.

There is one more place where we use `TypeMirror#hashCode()`:
 * [SelectionParameters#L72](https://github.com/mapstruct/mapstruct/blob/master/processor/src/main/java/org/mapstruct/ap/internal/model/source/SelectionParameters.java#L72)

There are other places where we use `TypeMirror#equals`:

* [SelectionParameters#L97](https://github.com/mapstruct/mapstruct/blob/master/processor/src/main/java/org/mapstruct/ap/internal/model/source/SelectionParameters.java#L97)
* [Type#L570](https://github.com/mapstruct/mapstruct/blob/master/processor/src/main/java/org/mapstruct/ap/internal/model/common/Type.java#L570)
* [MethodMatcher#L398](https://github.com/mapstruct/mapstruct/blob/master/processor/src/main/java/org/mapstruct/ap/internal/model/source/MethodMatcher.java#L398)

### PR DESCRIPTION
**Pull request by filiphr**
Fixes #1129

### CODE REVIEW

File: processor/src/main/java/org/mapstruct/ap/internal/model/EnumMappingMethod.java [hunk 1]
```diff
@@ -99,7 +100,7 @@ else if ( mappedConstants.size() == 1 ) {
                 }
             }
 
-            SelectionParameters selectionParameters = getSelecionParameters( method );
+            SelectionParameters selectionParameters = getSelecionParameters( method, ctx.getTypeUtils() );
```
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/internal/model/EnumMappingMethod.java]: selecionParameters = selectionParameters

File: processor/src/main/java/org/mapstruct/ap/internal/model/common/Type.java [hunk 2]
```diff
@@ -572,7 +572,7 @@ private Accessor getAdderForType(Type collectionProperty, String pluralPropertyN
                         continue;
                     }
                     VariableElement arg = executable.getParameters().get( 0 );
-                    if ( arg.asType().equals( typeArg ) ) {
+                    if ( typeUtils.isSameType( arg.asType(), typeArg ) ) {
```
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/internal/model/common/Type.java]: the `equals` method already uses `isSameType`..

Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/model/common/Type.java]: `arg.asType()` is a `TypeMirror` and not our `Type`. `arg` is also not our object but  `javax.lang.model.element.VariableElement`

File: processor/src/main/java/org/mapstruct/ap/internal/model/EnumMappingMethod.java (hunk 1 above)
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/model/EnumMappingMethod.java]: I can change that as well, but with #1197 this class will go away. Right?

Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/internal/model/EnumMappingMethod.java]: ok

File: processor/src/main/java/org/mapstruct/ap/internal/model/common/Type.java (hunk 2 above)
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/internal/model/common/Type.java]: OK, but isn't it strange that we use `TypeMirror` from other places iso `Type`

Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/model/common/Type.java]: This is only here locally, we need to get the type of the `ExecutableElement`

File: processor/src/main/java/org/mapstruct/ap/internal/model/source/SelectionParameters.java [hunk 3]
```diff
@@ -67,9 +71,8 @@ public TypeMirror getResultType() {
     @Override
     public int hashCode() {
         int hash = 3;
-        hash = 97 * hash + (this.qualifiers != null ? this.qualifiers.hashCode() : 0);
         hash = 97 * hash + (this.qualifyingNames != null ? this.qualifyingNames.hashCode() : 0);
-        hash = 97 * hash + (this.resultType != null ? this.resultType.hashCode() : 0);
+        hash = 97 * hash + (this.resultType != null ? this.resultType.toString().hashCode() : 0);
```
Comment (gunnarmorling) [inline processor/src/main/java/org/mapstruct/ap/internal/model/source/SelectionParameters.java]: Why is it using the name here?

**Comment by gunnarmorling** [comment]
Looks good. Should we add a rule in the build that forbids any calls of `TypeMirror#equals()`?  We could use the Forbidden APIs plug-in (not sure whether we already use that one or not).

File: processor/src/main/java/org/mapstruct/ap/internal/model/source/SelectionParameters.java (hunk 3 above)
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/model/source/SelectionParameters.java]: Which name do you mean? The `hashCode` is also not deterministic for `TypeMirror` and the `resultType` is a `TypeMirror`

**Comment by gunnarmorling** [comment]
By "name" I meant the toString() invocation. But it seems alright.

**Comment by filiphr** [comment]
>Looks good. Should we add a rule in the build that forbids any calls of TypeMirror#equals()? We could use the Forbidden APIs plug-in (not sure whether we already use that one or not).

Thanks. I think a rule would be good to have. Are you talking about [this](https://github.com/policeman-tools/forbidden-apis) plugin? I haven't used a plugin like this. I can try to set it up.

I mentioned in the issue about using sonarqube. With sonarqube we can hook it in with the PR. However, in build it might be better.

**Comment by gunnarmorling** [comment]
Yes, that's the one.

> However, in build it might be better.

Right, that way we'll notice about any such issues right when building
locally.

**Pull request merged**